mqtt-v5-broker/src/client.rs

```diff
@@ -20,28 +20,161 @@ use tokio::{
+                // to decode a new MQTT packet
+                loop {
+                    // Try to read an MQTT packet from the read buffer
+                    match mqtt_v5::decoder::decode_mqtt(&mut read_buf, protocol_version) {
+                        Ok(Some(packet)) => {
+                            if let Packet::Connect(packet) = &packet {
+                                protocol_version = packet.protocol_version;
+                            }
+
+                            // If we got one, return it
+                            return Some((Ok(packet), (ws_stream, read_buf, protocol_version)));
+                        },
+                        Err(e) => {
+                            // If we had a decode error, propagate the error along the stream
+                            return Some((Err(e), (ws_stream, read_buf, protocol_version)));
+                        },
+                        Ok(None) => {
+                            // Otherwise we need more binary data from the WebSocket stream
+                        },
+                    }
+
+                    let ws_frame = ws_stream.next().await;
+
+                    match ws_frame {
+                        Some(Ok(message)) => {
+                            if message.opcode() == Opcode::Close {
+                                return None;
+                            }
+
+                            if message.opcode() == Opcode::Ping {
+                                trace!("Got a websocket ping");
+                            }
+
+                            if message.opcode() != Opcode::Binary {
+                                // MQTT Control Packets MUST be sent in WebSocket binary data frames
+                                return Some((
+                                    Err(DecodeError::BadTransport),
+                                    (ws_stream, read_buf, protocol_version),
+                                ));
+                            }
+
+                            read_buf.extend_from_slice(&message.into_data());
+                        },
+                        Some(Err(e)) => {
+                            debug!("Error while reading from WebSocket stream: {:?}", e);
+                            // If we had a decode error in the WebSocket layer,
+                            // propagate the it along the stream
+                            return Some((
+                                Err(DecodeError::BadTransport),
+                                (ws_stream, read_buf, protocol_version),
+                            ));
+                        },
+                        None => {
+                            // The WebSocket stream is over, so we are too
+                            return None;
+                        },
+                    }
+                }
+            }
+        },
+    );
+
+    spawn_framed(Box::pin(stream), Box::pin(sink), broker_tx);
```

Comment:
So `Box::pin()` works with the heap and `tokio::pin!` works on the stack, is there a particular reason to prefer `Box::pin()` here? This is only out of curiosity because I have no idea if it makes a difference in practice.

Reply:
`tokio:pin!` doesn't provide `'static` because it expands to something on stack, while the `Box` can be moved into the task which makes `ST` fulfilling `'static` needed for `task::spawn`.
The two task are nice when looking at debugging profiles with `tokio::console` in order to see the load from the both pathes without any instrumentation. Performance wise there's probably more (and bigger) low hanging fruits like the clones of the `String` `client_id`.

I can revert the "two tasks` thing if requested. It's not needed but I couldn't resist ;-)

Reply:
Got it! I haven't used `tokio::console` before but if this style makes it easier to debug or measure performance then I'm cool with that. And I agree that performance is most likely not an issue.